File: ML/src/ppiq_ml/runtime/job_spec.py
"""The job specification the .NET runner writes and the Python runtime reads."""

from __future__ import annotations

import json
from dataclasses import dataclass, field, asdict
from typing import Any, Mapping

from .protocol import PROTOCOL_ID, ProtocolError, RefusalCode, check_protocol


@dataclass(frozen=True)
class ResourceBudget:
    """What the job is allowed to consume. Enforced by the caller, declared here."""

    max_wall_clock_seconds: float
    max_memory_mb: int = 0
    gpu_required: bool = False


@dataclass(frozen=True)
class ArtifactRef:
    """A sealed input artifact. The runtime never reads a database; it reads these."""

    artifact_id: str
    uri: str
    content_hash: str
    artifact_format: str
    byte_size: int = 0


@dataclass(frozen=True)
class JobSpec:
    """Identity, context and inputs for one ML execution.

    Carries no connection string, no table name and no SQL. The runtime's entire view
    of the customer's data is the sealed artifact list.
    """

    protocol: str
    job_id: str
    tenant_id: str
    site_id: str
    model_family: str
    inputs: tuple[ArtifactRef, ...]
    output_directory: str
    seed: int
    code_identity: str
    resources: ResourceBudget
    #: Present once the canonical Semantic Contract Manifest exists. Optional today,
    #: because SAFE-NOW work runs against fixture contracts.
    semantic_manifest_id: str | None = None
    checkpoint_directory: str | None = None
    cancellation_file: str | None = None
    parameters: Mapping[str, Any] = field(default_factory=dict)

    # ------------------------------------------------------------------ codec

    def to_json(self) -> str:
        return json.dumps(asdict(self), indent=2, sort_keys=True)

    @staticmethod
    def from_json(text: str) -> "JobSpec":
        try:
            raw = json.loads(text)
        except json.JSONDecodeError as exc:
            raise ProtocolError(
                RefusalCode.MALFORMED_JOB_SPEC,
                f"The job spec is not valid JSON: {exc}",
            ) from exc
        return JobSpec.from_dict(raw)

    @staticmethod
    def from_dict(raw: Mapping[str, Any]) -> "JobSpec":
        if not isinstance(raw, Mapping):
            raise ProtocolError(RefusalCode.MALFORMED_JOB_SPEC, "The job spec is not an object.")

        declared = raw.get("protocol")
        if declared is None:
            raise ProtocolError(
                RefusalCode.MALFORMED_JOB_SPEC,
                "The job spec declares no protocol. It was not interpreted.",
            )
        check_protocol(str(declared))

        required = (
            "job_id", "tenant_id", "site_id", "model_family",
            "output_directory", "seed", "code_identity", "resources",
        )
        missing = [k for k in required if k not in raw]
        if missing:
            raise ProtocolError(
                RefusalCode.MALFORMED_JOB_SPEC,
                f"The job spec is missing required fields: {', '.join(sorted(missing))}.",
            )

        res = raw["resources"]
        if not isinstance(res, Mapping) or "max_wall_clock_seconds" not in res:
            raise ProtocolError(
                RefusalCode.MALFORMED_JOB_SPEC,
                "The job spec declares no wall-clock budget.",
            )

        inputs = tuple(
            ArtifactRef(
                artifact_id=str(a["artifact_id"]),
                uri=str(a["uri"]),
                content_hash=str(a["content_hash"]),
                artifact_format=str(a["artifact_format"]),
                byte_size=int(a.get("byte_size", 0)),
            )
            for a in raw.get("inputs", [])
        )

        return JobSpec(
            protocol=PROTOCOL_ID,
            job_id=str(raw["job_id"]),
            tenant_id=str(raw["tenant_id"]),
            site_id=str(raw["site_id"]),
            model_family=str(raw["model_family"]),
            inputs=inputs,
            output_directory=str(raw["output_directory"]),
            seed=int(raw["seed"]),
            code_identity=str(raw["code_identity"]),
            resources=ResourceBudget(
                max_wall_clock_seconds=float(res["max_wall_clock_seconds"]),
                max_memory_mb=int(res.get("max_memory_mb", 0)),
                gpu_required=bool(res.get("gpu_required", False)),
            ),
            semantic_manifest_id=(
                str(raw["semantic_manifest_id"]) if raw.get("semantic_manifest_id") else None
            ),
            checkpoint_directory=(
                str(raw["checkpoint_directory"]) if raw.get("checkpoint_directory") else None
            ),
            cancellation_file=(
                str(raw["cancellation_file"]) if raw.get("cancellation_file") else None
            ),
            parameters=dict(raw.get("parameters", {})),
        )

File: Backend/PlantProcess.ML.Runtime/JobSpec.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace PlantProcess.ML.Runtime;

/// <summary>What the job is allowed to consume. Declared here, enforced by this side.</summary>
public sealed record ResourceBudget(
    [property: JsonPropertyName("max_wall_clock_seconds")] double MaxWallClockSeconds,
    [property: JsonPropertyName("max_memory_mb")] int MaxMemoryMb = 0,
    [property: JsonPropertyName("gpu_required")] bool GpuRequired = false);

/// <summary>
/// A sealed input artifact. The Python runtime never reads a database; it reads these.
/// A physical schema change is absorbed by the snapshot materialiser, not by a model.
/// </summary>
public sealed record ArtifactRef(
    [property: JsonPropertyName("artifact_id")] string ArtifactId,
    [property: JsonPropertyName("uri")] string Uri,
    [property: JsonPropertyName("content_hash")] string ContentHash,
    [property: JsonPropertyName("artifact_format")] string ArtifactFormat,
    [property: JsonPropertyName("byte_size")] long ByteSize = 0);

/// <summary>
/// Identity, context and inputs for one ML execution. Carries no connection string,
/// no table name and no SQL.
/// </summary>
public sealed record JobSpec
{
    [JsonPropertyName("protocol")] public string Protocol { get; init; } = MlJobProtocol.Id;
    [JsonPropertyName("job_id")] public string JobId { get; init; } = string.Empty;
    [JsonPropertyName("tenant_id")] public string TenantId { get; init; } = string.Empty;
    [JsonPropertyName("site_id")] public string SiteId { get; init; } = string.Empty;
    [JsonPropertyName("model_family")] public string ModelFamily { get; init; } = string.Empty;
    [JsonPropertyName("inputs")] public IReadOnlyList<ArtifactRef> Inputs { get; init; } = Array.Empty<ArtifactRef>();
    [JsonPropertyName("output_directory")] public string OutputDirectory { get; init; } = string.Empty;
    [JsonPropertyName("seed")] public int Seed { get; init; }
    [JsonPropertyName("code_identity")] public string CodeIdentity { get; init; } = string.Empty;
    [JsonPropertyName("resources")] public ResourceBudget Resources { get; init; } = new(0);

    /// <summary>Present once the canonical Semantic Contract Manifest exists. Optional today.</summary>
    [JsonPropertyName("semantic_manifest_id")] public string? SemanticManifestId { get; init; }

    [JsonPropertyName("checkpoint_directory")] public string? CheckpointDirectory { get; init; }
    [JsonPropertyName("cancellation_file")] public string? CancellationFile { get; init; }
    [JsonPropertyName("parameters")] public IReadOnlyDictionary<string, JsonElement> Parameters { get; init; }
        = new Dictionary<string, JsonElement>();

    public static readonly JsonSerializerOptions Json = new()
    {
        WriteIndented = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.Never
    };

    public string ToJson() => JsonSerializer.Serialize(this, Json);

    public static JobSpec FromJson(string text)
    {
        JobSpec? spec;
        try
        {
            spec = JsonSerializer.Deserialize<JobSpec>(text, Json);
        }
        catch (JsonException exception)
        {
            throw new MlProtocolException(MlRefusalCode.MalformedJobSpec,
                $"The job spec is not valid JSON: {exception.Message}");
        }

        if (spec is null)
            throw new MlProtocolException(MlRefusalCode.MalformedJobSpec, "The job spec is empty.");

        if (string.IsNullOrWhiteSpace(spec.Protocol))
            throw new MlProtocolException(MlRefusalCode.MalformedJobSpec,
                "The job spec declares no protocol. It was not interpreted.");

        if (spec.Protocol != MlJobProtocol.Id)
            throw new MlProtocolException(MlRefusalCode.ProtocolVersionMismatch,
                $"Job spec declares protocol '{spec.Protocol}'; this runtime speaks "
                + $"'{MlJobProtocol.Id}'. The payload was not interpreted.");

        var missing = new List<string>();
        if (string.IsNullOrWhiteSpace(spec.JobId)) missing.Add("job_id");
        if (string.IsNullOrWhiteSpace(spec.TenantId)) missing.Add("tenant_id");
        if (string.IsNullOrWhiteSpace(spec.SiteId)) missing.Add("site_id");
        if (string.IsNullOrWhiteSpace(spec.ModelFamily)) missing.Add("model_family");
        if (string.IsNullOrWhiteSpace(spec.OutputDirectory)) missing.Add("output_directory");
        if (string.IsNullOrWhiteSpace(spec.CodeIdentity)) missing.Add("code_identity");
        if (missing.Count > 0)
            throw new MlProtocolException(MlRefusalCode.MalformedJobSpec,
                $"The job spec is missing required fields: {string.Join(", ", missing.OrderBy(m => m))}.");

        if (spec.Resources is null || spec.Resources.MaxWallClockSeconds <= 0)
            throw new MlProtocolException(MlRefusalCode.MalformedJobSpec,
                "The job spec declares no wall-clock budget.");

        return spec;
    }
}
